File: src-tauri/src/ai/llm_client.rs
use serde::{Deserialize, Serialize};
use crate::error::{AppError, AppResult};

/// LLM 消息
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub role: String,
    pub content: String,
}

/// LLM 调用选项
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LLMOptions {
    pub temperature: Option<f64>,
    pub max_tokens: Option<i32>,
    pub top_p: Option<f64>,
}

impl Default for LLMOptions {
    fn default() -> Self {
        Self {
            temperature: Some(0.7),
            max_tokens: Some(2048),
            top_p: Some(0.9),
        }
    }
}

/// 硅基流动 LLM 客户端
#[derive(Clone)]
pub struct LLMClient {
    api_base: String,
    api_key: String,
    model: String,
    client: reqwest::Client,
}

/// OpenAI 兼容响应结构
#[derive(Debug, Deserialize)]
struct ChatCompletionResponse {
    choices: Vec<Choice>,
}

#[derive(Debug, Deserialize)]
struct Choice {
    message: ResponseMessage,
}

#[derive(Debug, Deserialize)]
struct ResponseMessage {
    content: String,
}

/// 流式 chunk 结构
#[derive(Debug, Deserialize)]
struct StreamChunk {
    choices: Vec<StreamChoice>,
}

#[derive(Debug, Deserialize)]
struct StreamChoice {
    delta: StreamDelta,
    #[serde(default)]
    finish_reason: Option<String>,
}

#[derive(Debug, Deserialize)]
struct StreamDelta {
    #[serde(default)]
    content: Option<String>,
}

impl LLMClient {
    /// 创建新的 LLM 客户端
    pub fn new(api_base: &str, api_key: &str, model: &str) -> Self {
        Self {
            api_base: api_base.to_string(),
            api_key: api_key.to_string(),
            model: model.to_string(),
            client: reqwest::Client::new(),
        }
    }

    /// 普通完成（等待完整响应）
    pub async fn complete(&self, messages: &[Message], opts: &LLMOptions) -> AppResult<String> {
        let url = format!("{}/chat/completions", self.api_base);

        let body = serde_json::json!({
            "model": self.model,
            "messages": messages,
            "temperature": opts.temperature.unwrap_or(0.7),
            "max_tokens": opts.max_tokens.unwrap_or(2048),
            "top_p": opts.top_p.unwrap_or(0.9),
            "stream": false
        });

        let response = self.client
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| AppError::LLMError(format!("请求失败: {}", e)))?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            return Err(AppError::LLMError(format!("API 返回 {}: {}", status, text)));
        }

        let result: ChatCompletionResponse = response.json().await
            .map_err(|e| AppError::LLMError(format!("解析响应失败: {}", e)))?;

        result.choices.first()
            .map(|c| c.message.content.clone())
            .ok_or_else(|| AppError::LLMError("响应中无 choices".to_string()))
    }

    /// 流式完成 — 通过回调函数逐 token 输出
    pub async fn complete_stream<F>(
        &self,
        messages: &[Message],
        opts: &LLMOptions,
        mut on_chunk: F,
    ) -> AppResult<String>
    where
        F: FnMut(&str),
    {
        let url = format!("{}/chat/completions", self.api_base);

        let body = serde_json::json!({
            "model": self.model,
            "messages": messages,
            "temperature": opts.temperature.unwrap_or(0.7),
            "max_tokens": opts.max_tokens.unwrap_or(2048),
            "top_p": opts.top_p.unwrap_or(0.9),
            "stream": true
        });

        let response = self.client
            .post(&url)
            .header("Authorization", format!("Bearer {}", self.api_key))
            .header("Content-Type", "application/json")
            .json(&body)
            .send()
            .await
            .map_err(|e| AppError::LLMError(format!("请求失败: {}", e)))?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            return Err(AppError::LLMError(format!("API 返回 {}: {}", status, text)));
        }

        let mut full_content = String::new();
        let mut buffer = String::new();

        // 使用 bytes_stream 逐块读取
        use futures::StreamExt;
        let mut stream = response.bytes_stream();

        while let Some(chunk_result) = stream.next().await {
            let chunk = chunk_result
                .map_err(|e| AppError::LLMError(format!("读取流失败: {}", e)))?;

            let text = String::from_utf8_lossy(&chunk);
            buffer.push_str(&text);

            // SSE 格式：每行以 "data: " 开头
            while let Some(line_end) = buffer.find('\n') {
                let line = buffer[..line_end].trim().to_string();
                buffer = buffer[line_end + 1..].to_string();

                if line.is_empty() || line == "data: [DONE]" {
                    continue;
                }

                if let Some(json_str) = line.strip_prefix("data: ") {
                    if let Ok(chunk) = serde_json::from_str::<StreamChunk>(json_str) {
                        for choice in &chunk.choices {
                            if let Some(ref content) = choice.delta.content {
                                full_content.push_str(content);
                                on_chunk(content);
                            }
                        }
                    }
                }
            }
        }

        Ok(full_content)
    }

    /// 切换模型
    pub fn set_model(&mut self, model_id: &str) {
        self.model = model_id.to_string();
        tracing::info!("LLM 模型切换为: {}", model_id);
    }

    /// 获取当前模型
    pub fn current_model(&self) -> &str {
        &self.model
    }
}

File: server/app/ai/llm_client.py
"""LLM 客户端 - 从 src-tauri/src/ai/llm_client.rs 移植

使用 httpx (AsyncClient) 调硅基流动 OpenAI 兼容 API。
提供普通 complete 和流式 complete_stream 两种接口。
"""
import json
import logging
from collections.abc import AsyncIterator
from dataclasses import dataclass

import httpx

logger = logging.getLogger(__name__)


@dataclass
class Message:
    role: str  # 'system' | 'user' | 'assistant'
    content: str

    def to_dict(self) -> dict:
        return {"role": self.role, "content": self.content}


@dataclass
class LLMOptions:
    temperature: float = 0.7
    max_tokens: int = 2048
    top_p: float = 0.9


class LLMError(Exception):
    pass


class LLMClient:
    """硅基流动 LLM 客户端 (OpenAI 兼容)"""

    def __init__(self, api_base: str, api_key: str, model: str):
        self.api_base = api_base.rstrip("/")
        self.api_key = api_key
        self.model = model

    def set_model(self, model: str) -> None:
        self.model = model
        logger.info(f"LLM 模型切换为: {model}")

    def _headers(self) -> dict[str, str]:
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

    def _body(self, messages: list[Message], opts: LLMOptions, stream: bool) -> dict:
        return {
            "model": self.model,
            "messages": [m.to_dict() for m in messages],
            "temperature": opts.temperature,
            "max_tokens": opts.max_tokens,
            "top_p": opts.top_p,
            "stream": stream,
        }

    async def complete(
        self,
        messages: list[Message],
        opts: LLMOptions | None = None,
    ) -> str:
        """普通完成 - 等待完整响应"""
        opts = opts or LLMOptions()
        url = f"{self.api_base}/chat/completions"
        body = self._body(messages, opts, stream=False)

        async with httpx.AsyncClient(timeout=60.0) as client:
            try:
                response = await client.post(url, headers=self._headers(), json=body)
            except httpx.RequestError as e:
                raise LLMError(f"请求失败: {e}") from e

            if response.status_code != 200:
                raise LLMError(f"API 返回 {response.status_code}: {response.text}")

            data = response.json()

        try:
            return data["choices"][0]["message"]["content"]
        except (KeyError, IndexError) as e:
            raise LLMError(f"响应中无 choices: {data}") from e

    async def complete_stream(
        self,
        messages: list[Message],
        opts: LLMOptions | None = None,
    ) -> AsyncIterator[str]:
        """流式完成 - 逐 token 异步迭代输出"""
        opts = opts or LLMOptions()
        url = f"{self.api_base}/chat/completions"
        body = self._body(messages, opts, stream=True)

        async with httpx.AsyncClient(timeout=60.0) as client:
            async with client.stream(
                "POST", url, headers=self._headers(), json=body
            ) as response:
                if response.status_code != 200:
                    text = await response.aread()
                    raise LLMError(
                        f"API 返回 {response.status_code}: {text.decode('utf-8', errors='replace')}"
                    )

                buffer = ""
                async for chunk in response.aiter_text():
                    buffer += chunk
                    while "\n" in buffer:
                        line, buffer = buffer.split("\n", 1)
                        line = line.strip()
                        if not line or line == "data: [DONE]":
                            continue
                        if not line.startswith("data: "):
                            continue
                        try:
                            payload = json.loads(line[6:])
                        except json.JSONDecodeError:
                            continue
                        choices = payload.get("choices", [])
                        if not choices:
                            continue
                        delta = choices[0].get("delta", {})
                        content = delta.get("content")
                        if content:
                            yield content


# 全局单例（由 main.py 或 config 初始化）
_global_client: LLMClient | None = None


def get_llm_client() -> LLMClient:
    """按需从 settings 构造全局 LLMClient"""
    global _global_client
    if _global_client is None:
        from ..config import get_settings

        s = get_settings()
        _global_client = LLMClient(
            api_base=s.siliconflow_api_base,
            api_key=s.siliconflow_api_key,
            model=s.siliconflow_default_model,
        )
    return _global_client
